REGRESSION( iOS 26): 2X TestWebKitAPI.ProcessSwap gesture controller API Tests are constant failures

colelao opened the pull request:
#### 49c4b8c1c84e37bc1a11d4c53f6a144c07f76475
<pre>
REGRESSION( iOS 26): 2X TestWebKitAPI.ProcessSwap gesture controller API Tests are constant failures
<a href="https://bugs.webkit.org/show_bug.cgi?id=301536">https://bugs.webkit.org/show_bug.cgi?id=301536</a>
<a href="https://rdar.apple.com/163517689">rdar://163517689</a>

Reviewed by Aditya Keerthi and Rupin Mittal.

The gesture recognizers were not being properly removed on process swaps. I added a
new method platformDisconnectFromProcess to remove the stale gesture recognizers.

Test: Tools/TestWebKitAPI/Tests/WebKitCocoa/ProcessSwapOnNavigation.mm

* Source/WebKit/UIProcess/ViewGestureController.cpp:
(WebKit::ViewGestureController::disconnectFromProcess):
* Source/WebKit/UIProcess/ViewGestureController.h:
* Source/WebKit/UIProcess/gtk/ViewGestureControllerGtk.cpp:
(WebKit::ViewGestureController::platformDisconnectFromProcess):
* Source/WebKit/UIProcess/ios/ViewGestureControllerIOS.mm:
(-[WKSwipeTransitionController invalidate]):
(WebKit::ViewGestureController::platformDisconnectFromProcess):
* Source/WebKit/UIProcess/mac/ViewGestureControllerMac.mm:
(WebKit::ViewGestureController::platformDisconnectFromProcess):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/ProcessSwapOnNavigation.mm:
((ProcessSwap, SwapWithGestureController)):
((ProcessSwap, CrashWithGestureController)(ProcessSwap, CrashWithGestureController)):
((ProcessSwap, DISABLED_SwapWithGestureController)(ProcessSwap, SwapWithGestureController)): Deleted.
((ProcessSwap, DISABLED_CrashWithGestureController)(ProcessSwap, CrashWithGestureController)): Deleted.

Canonical link: <a href="https://commits.webkit.org/304570@main">https://commits.webkit.org/304570@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/34be213b80332a7ec11da9ccaa3162d5793af1c8

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/135915 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/8286 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/47201 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/143626 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/87474 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/850eb649-b898-461f-ba0f-0c1b8929b21f) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/137784 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/8945 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/8130 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/103871 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/71776 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/d1711873-d55d-44a3-b37d-6fd3c71d4f3e) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/138861 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/6469 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/121814 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/84748 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/ae464459-bbc1-4261-8e03-28d948d312bb) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/6174 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/3818 "Passed tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/4230 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/115439 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/40002 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/146375 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/7969 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/40597 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/112225 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/7991 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/6689 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/112614 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28587 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/6082 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/118112 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/61867 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/8016 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/36191 "Passed tests") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/7745 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/71568 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/7965 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/7827 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

RupinMittal commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/ProcessSwapOnNavigation.mm:
```diff
@@ -6247,11 +6247,7 @@ static bool viewHasSwipeGestures(UIView *view)
 #endif
 
 // rdar://163517689 (REGRESSION( iOS 26): 2X TestWebKitAPI.ProcessSwap (API-Tests) are constant failures (301536))
```
Now that this is being fixed, we should remove this line.

RupinMittal commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/ProcessSwapOnNavigation.mm:
```diff
@@ -6292,11 +6288,7 @@ static bool viewHasSwipeGestures(UIView *view)
 }
 
 // rdar://163517689 (REGRESSION( iOS 26): 2X TestWebKitAPI.ProcessSwap (API-Tests) are constant failures (301536))
```
And this one too.

RupinMittal reviewed (approved):
Thanks for the fix! LGTM with small comments. It would be good to get a review from @hortont424 or @pxlcoder too since they're more familiar with this code than me.

jesxilin commented on Source/WebKit/UIProcess/ViewGestureController.h:
```diff
@@ -115,6 +115,7 @@ class ViewGestureController final : public IPC::MessageReceiver, public RefCount
     void deref() const final { RefCounted::deref(); }
 
     void platformTeardown();
+    void platformDisconnectFromProcess();
```
Nit Suggestion: It seems like `platformDisconnectFromProcess()` is only used/implemented for iOS rn. I think alternatively, you could guard this with `#if PLATFORM(IOS_FAMILY)` and delete the other functions.

jesxilin commented on Source/WebKit/UIProcess/ios/ViewGestureControllerIOS.mm:
```diff
@@ -88,6 +88,13 @@ - (instancetype)initWithViewGestureController:(WebKit::ViewGestureController*)ge
 - (void)invalidate
 {
+    if (_gestureRecognizerView) {
+        if (UIGestureRecognizer *backRecognizer = [_backTransitionController gestureRecognizer])
```
You should use RetainPtr instead of the raw ptr. 
`if (RetainPtr backRecognizer = [_backTransitionController gestureRecognizer])`

jesxilin commented on Source/WebKit/UIProcess/ios/ViewGestureControllerIOS.mm:
```diff
@@ -88,6 +88,13 @@ - (instancetype)initWithViewGestureController:(WebKit::ViewGestureController*)ge
+    if (_gestureRecognizerView) {
+        if (UIGestureRecognizer *backRecognizer = [_backTransitionController gestureRecognizer])
+            [_gestureRecognizerView.get() removeGestureRecognizer:backRecognizer];
+        if (UIGestureRecognizer *forwardRecognizer = [_forwardTransitionController gestureRecognizer])
```
Ditto.

jesxilin commented on Source/WebKit/UIProcess/ios/ViewGestureControllerIOS.mm:
```diff
@@ -88,6 +88,13 @@ - (instancetype)initWithViewGestureController:(WebKit::ViewGestureController*)ge
 {
+    if (_gestureRecognizerView) {
+        if (UIGestureRecognizer *backRecognizer = [_backTransitionController gestureRecognizer])
+            [_gestureRecognizerView.get() removeGestureRecognizer:backRecognizer];
```
Then this would be `[_gestureRecognizerView removeGestureRecognizer:backRecognizer.get()];`

pxlcoder reviewed (approved):
Looks good with @jesxilin's comments addressed. I don't feel strongly about making `platformDisconnectFromProcess` iOS only though.

colelao committed:
REGRESSION( iOS 26): 2X TestWebKitAPI.ProcessSwap gesture controller API Tests are constant failures
https://bugs.webkit.org/show_bug.cgi?id=301536
rdar://163517689

Reviewed by Aditya Keerthi and Rupin Mittal.

The gesture recognizers were not being properly removed on process swaps. I added a
new method platformDisconnectFromProcess to remove the stale gesture recognizers.

Test: Tools/TestWebKitAPI/Tests/WebKitCocoa/ProcessSwapOnNavigation.mm

* Source/WebKit/UIProcess/ViewGestureController.cpp:
(WebKit::ViewGestureController::disconnectFromProcess):
* Source/WebKit/UIProcess/ViewGestureController.h:
* Source/WebKit/UIProcess/gtk/ViewGestureControllerGtk.cpp:
(WebKit::ViewGestureController::platformDisconnectFromProcess):
* Source/WebKit/UIProcess/ios/ViewGestureControllerIOS.mm:
(-[WKSwipeTransitionController invalidate]):
(WebKit::ViewGestureController::platformDisconnectFromProcess):
* Source/WebKit/UIProcess/mac/ViewGestureControllerMac.mm:
(WebKit::ViewGestureController::platformDisconnectFromProcess):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/ProcessSwapOnNavigation.mm:
((ProcessSwap, SwapWithGestureController)):
((ProcessSwap, CrashWithGestureController)(ProcessSwap, CrashWithGestureController)):
((ProcessSwap, DISABLED_SwapWithGestureController)(ProcessSwap, SwapWithGestureController)): Deleted.
((ProcessSwap, DISABLED_CrashWithGestureController)(ProcessSwap, CrashWithGestureController)): Deleted.

Canonical link: https://commits.webkit.org/304570@main

pxlcoder commented on Source/WebKit/UIProcess/ios/ViewGestureControllerIOS.mm:
```diff
@@ -88,6 +88,13 @@ - (instancetype)initWithViewGestureController:(WebKit::ViewGestureController*)ge
 {
+    if (_gestureRecognizerView) {
+        if (RetainPtr backRecognizer = [_backTransitionController gestureRecognizer])
+            [_gestureRecognizerView.get() removeGestureRecognizer:backRecognizer.get()];
+        if (RetainPtr forwardRecognizer = [_forwardTransitionController gestureRecognizer])
+            [_gestureRecognizerView.get() removeGestureRecognizer:forwardRecognizer.get()];
```
It looks like this comment was missed: https://github.com/WebKit/WebKit/pull/55345#discussion_r2624525707

There's no need for `.get()` on `_gestureRecognizerView`. A method call on `RetainPtr` will work on the underlying pointer.

Not urgent, but could clean in a follow-up PR.
